### DISCUSSION TITLE
True suppression of errors or "Am I holding try_ wrong?"

### DISCUSSION THREAD

**Discussion by jevinskie** [foonathan/lexy#117] [Q&A] [answered]
I added the `else_` to make a branch-rule out of a rule. Is that at all sensible?

In this example I'd like to match either `hello world` or `hello :: world` and not produce an error since it still makes a successful production and I'd like `is_error` to represent that. Am I on the right path or do I need a new approach to branches, conditions, and errors?

https://lexy.foonathan.net/playground/?id=KsjjTseno&mode=tree

**Comment by foonathan** [comment, marked as answer]
So I think there is a bit of a misunderstanding with `dsl::try_`: it is only used for error recovery. If you have a situation where a token is required by the grammar, but parsing can continue if it's omitted, you can use `dsl::try_` for that. The error will still be reported, but instead of canceling the production, parsing can continue to check the rest of the input.

Also, `dsl::opt(dsl::else_ >> x)` is never correct, as that is equivalent to `x`: `dsl::else_` is a branch rule that is unconditionally taken and is meant in combination with `|`.

I didn't quite understand what you want, so here are possible interpretations:

1. You want to parse `hello world` or `hello::world`, where either is correct. Use `dsl::opt()` (or `dsl::if_()`) with `LEXY_LIT("::")` (`dsl::colon + dsl::colon` isn't a branch rule): https://lexy.foonathan.net/playground/?id=1dz7vEofc&mode=tree
2. You want to parse `hello::world` but recover when the user omitted the `::`. This is where `dsl::try_()` comes into a play, as a missing `::` is still an error: https://lexy.foonathan.net/playground/?id=GEKh3zMex&mode=tree Note that a single `:` confuses it, since recovery just skips parsing the `::` and continues with parsing `world`, which fails as there is trailing whitespace on the input: https://lexy.foonathan.net/playground/?id=43G5df3KG&mode=trace
3. You want to parse `hello::world` but recover when the user omitted the `::` or used a single `:` instead. You can either use `dsl::try_(dsl::colon) + dsl::try_(dsl::colon)`, or use a special recovery rule to skip trailing whitespace: https://lexy.foonathan.net/playground/?id=Gvxz8ax4j&mode=trace

(I've just now noticed that there is `dsl::double_colon` as well, which is a shorthand for `LEXY_LIT("::")`).

I should probably make the behavior in 3 the default, i.e. skip whitespace after error recovery; I haven't considered that situation before.

**Reply by foonathan** [reply]
> I should probably make the behavior in 3 the default, i.e. skip whitespace after error recovery; I haven't considered that situation before.

This has been implemented in 30eb8008.
